internal/topology/check/compatibility.go

```diff
@@ -54,3 +62,261 @@ func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
+			current.Ref.GetObjectKind().GroupVersionKind(), current.Ref.Name, currentGK, desiredGK)
+	}
+	return TemplatesAreInTheSameNamespace(current, desired)
+}
+
+// TemplatesAreInTheSameNamespace checks if two referenced objects are in the same namespace.
+func TemplatesAreInTheSameNamespace(current, desired clusterv1.LocalObjectTemplate) error {
+	if current.Ref.Namespace != desired.Ref.Namespace {
+		return errors.Errorf("invalid operation: it is not possible to change the namespace of %s/%s from %s to %s",
+			current.Ref.GetObjectKind().GroupVersionKind(), current.Ref.Name, current.Ref.Namespace, desired.Ref.Namespace)
+	}
+	return nil
+}
+
+// TemplateIsValid ensures the template is in the correct namespace, has no nil references, and has a valid Kind and GroupVersion.
+func TemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace string, pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// check if ref is not nil.
+	if in.Ref == nil {
+		return field.ErrorList{field.Invalid(
+			pathPrefix.Child("ref"),
+			in.Ref.Name,
+			"cannot be nil",
+		)}
+	}
+
+	// check if a name is provided
+	if in.Ref.Name == "" {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "name"),
+				in.Ref.Name,
+				"cannot be empty",
+			),
+		)
+	}
+
+	// validate if namespace matches the provided namespace
+	if namespace != "" && in.Ref.Namespace != namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "namespace"),
+				in.Ref.Namespace,
+				fmt.Sprintf("must be '%s'", namespace),
+			),
+		)
+	}
+
+	// check if kind is a template
+	if len(in.Ref.Kind) <= len(clusterv1.TemplateSuffix) || !strings.HasSuffix(in.Ref.Kind, clusterv1.TemplateSuffix) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "kind"),
+				in.Ref.Kind,
+				fmt.Sprintf("kind must be of form '<name>%s'", clusterv1.TemplateSuffix),
+			),
+		)
+	}
+
+	// check if apiVersion is valid
+	gv, err := schema.ParseGroupVersion(in.Ref.APIVersion)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "apiVersion"),
+				in.Ref.APIVersion,
+				fmt.Sprintf("must be a valid apiVersion: %v", err),
+			),
+		)
+	}
+	if err == nil && gv.Empty() {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "apiVersion"),
+				in.Ref.APIVersion,
+				"value cannot be empty",
+			),
+		)
+	}
+	return allErrs
+}
+
+// ClusterClassesAreCompatible checks that compatibility between new and old versions of a Cluster Class.
+// It checks that:
+// 1) Infrastructure Templates are compatible.
+// 2) ControlPlane Templates are compatible.
+// 3) ControlPlane Machine Infrastructure Templates are compatible.
+// 4) MachineDeploymentClasses have not been deleted and are compatible.
+func ClusterClassesAreCompatible(old, in *clusterv1.ClusterClass) field.ErrorList {
```

Comment:
Can we try to avoid `in`? Maybe `old`/`new`?

The other funcs are using current/desired but not sure if that fits here good enough